File: abc/abc128/abc128_c.cs
// https://atcoder.jp/contests/abc128/submissions/8389684

using System;
using System.Linq;

namespace CSharp
{
    class Program
    {
        public static void Main()
        {
            var input = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();
            var N = input[0];
            var M = input[1];

            var s = new int[M][];
            for (int i = 0; i < M; i++)
            {
                s[i] = Console.ReadLine().Split(' ').Select(int.Parse).Skip(1).ToArray();
            }

            var p = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();

            var result = 0;
            for (int i = 0; i < Math.Pow(2, N); i++)
            {
                var c = Convert.ToString(i, 2).PadLeft(N, '0').ToCharArray();

                var flag = true;
                for (int j = 0; j < M; j++)
                {
                    var temp = 0;
                    for (int k = 0; k < s[j].Length; k++)
                    {
                        temp += c[s[j][k] - 1];
                    }
                    if (temp % 2 != p[j])
                    {
                        flag = false;
                        break;
                    }
                }
                if (flag)
                {
                    result++;
                }
            }

            Console.WriteLine(result);
        }
    }
}

File: abc/abc128/abc128_c.fs
// https://atcoder.jp/contests/abc128/submissions/12226417

open System

let [| N; M |] = stdin.ReadLine().Split() |> Array.map int

let s = Array.zeroCreate<int []> M

for i in [ 0..M - 1 ] do
    s.[i] <- stdin.ReadLine().Split()
             |> Array.skip 1
             |> Array.map int

let p = stdin.ReadLine().Split() |> Array.map int

let mutable result = 0

for i in [ 0..pown 2 N - 1 ] do
    let c = Convert.ToString(i, 2).PadLeft(N, '0').ToCharArray()
    let mutable flag = true
    for j in [ 0..M - 1 ] do
        if flag then
            let mutable t = 0
            for k in [ 0..(s.[j] |> Array.length) - 1 ] do
                t <- t + int c.[s.[j].[k] - 1]
            if t % 2 <> p.[j] then flag <- false
    if flag then result <- result + 1

result |> stdout.WriteLine
